src/panel/pages/timeline/components/TimelineTooltip.tsx

```diff
@@ -42,59 +55,103 @@ const TooltipElement = styled.p<{ position?: TooltipPosition }>`
     border-left: 6px solid transparent;
     border-right: 6px solid transparent;
     margin-top: -1px;
-    left: 50%;
+    left: calc(50% - ${(p) => p.offset}px);
     top: 100%;
     transform: translate(-50%, 0);
   }
 `;
 
 export const useTooltip = () => {
-  const ref: MutableRefObject<HTMLElement | null> = useRef<HTMLElement>(null);
+  const containerRef: MutableRefObject<HTMLElement | null> = useRef<
+    HTMLElement
+  >(null);
+  const tooltipRef: MutableRefObject<HTMLElement | null> = useRef<HTMLElement>(
+    null
+  );
+  const tooltipOffset: MutableRefObject<number> = useRef(0);
+  const mouseX = useRef<number | undefined>(undefined);
   const [hasRef, setHasRef] = useState(false);
   const [isVisible, setIsVisible] = useState(false);
   const [tooltipProps, setTooltipProps] = useState<
-    ComponentProps<typeof TimelineTooltip>
+    ComponentProps<typeof TimelineTooltip> & TimelineTooltipProps
   >({});
 
   const calculateTooltipPosition = useCallback(() => {
-    if (!ref.current) {
+    if (!containerRef.current) {
       return;
     }
 
-    const { x, y, width } = ref.current.getBoundingClientRect(); //eslint-disable-line @typescript-eslint/no-non-null-assertion
-    setTooltipProps({
+    const { x, y, width } = containerRef.current.getBoundingClientRect(); //eslint-disable-line @typescript-eslint/no-non-null-assertion
+
+    const newTooltipProps: ComponentProps<typeof TimelineTooltip> &
+      TimelineTooltipProps = {
       style: {
         position: "fixed",
         left: x + width / 2,
         top: y,
         transform: `translateX(-50%) translateY(-100%)`,
+        paddingBottom: 8,
+        zIndex: 9,
       },
+      handleTooltipRef,
+      tooltipOffset: tooltipOffset.current,
+    };
+
+    if (!tooltipRef.current || !!tooltipOffset.current) {
+      setTooltipProps(newTooltipProps);
+      return;
+    }
+
+    const offsetRight =
+      window.innerWidth - tooltipRef.current.getBoundingClientRect().right;
+    const offsetLeft = tooltipRef.current.getBoundingClientRect().left;
+
+    // * add an offset to the tooltip if it would be rendered out of bounds
+    if (offsetLeft < 0) {
+      tooltipOffset.current = Math.abs(offsetLeft) + 8;
+    }
+
+    if (offsetRight < 0) {
+      tooltipOffset.current = offsetRight - 8;
+    }
+
+    setTooltipProps({
+      ...newTooltipProps,
+      tooltipOffset: tooltipOffset.current,
     });
   }, []);
 
-  const handleRef = useCallback((passedRef) => {
+  const handleContainerRef = useCallback((passedRef) => {
```

Comment:
Nit on wording: If this is the hover target, would be useful to call it a _handleTargetRef_ to avoid confusion with the timeline containing element we use in context.